docs/decisions/broker-trigger.md

```diff
@@ -0,0 +1,166 @@
+# Broker and Trigger model (for 0.5.0)
+
+Decision date: 6 March 2019
+
+## Define High-Level Objects Related to Eventing
+
+The goal of eventing is to provide an abstract "bucket of events" which event
+consumers can sample from using filters on CloudEvent
+[context attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes).
+Event producers can deliver events into the bucket without needing to understand
+the exact mechanics of routing to specific consumers. Event producers and
+consumers are decoupled temporally but also through a deliberately black-box
+abstraction which can optimize the underlying routing layer dynamically.
+
+## Definitions
+
+- **Event Producer**: a system which creates events based on occurrences. E.g.
+  GitHub (the hosted service).
+- **Event Consumer**: a destination which receives events. E.g. the application
+  processing an event. Typically, this is the "sink" in the processing flow of
+  an Event, but could be a data processing system like Spark.
+- **Sink**: a generic term to mean the destination for an event being sent from
+  a component. Synonym for Event Consumer.
+- **Event Source**: a helper object that can be used to connect an Event
+  Producer to the Knative eventing system by routing events to an Event
+  Consumer. E.g. the
+  [GitHub Knative Source](https://github.com/knative/eventing-sources/tree/master/pkg/reconciler/githubsource)
+- **Receive Adapter**: a data plane entity (Knative Service, Deployment, etc)
```

Comment:
I agree with this definition because it implies to be its more like a proxy. It "performs the event routing". And in that scenario the consumer would expect things like the CE-source to be something related to the original event producer. However, I know this view is not universally shared. So, we need to decide... is a Receiver Adapter just a piece of middleware or does it become the Event Producer ? And if the answer is "it depends on the Event Source" then let's call that out so it's clear that event consumer can't assume one model for all Receive Adapters.

Reply:
I think you're actually thinking of the `type`, not `source` context attribute?

Some guidance from CloudEvents would be helpful here; I think currently CloudEvents describes middleware (including routers) as "Event Consumers" and then "Event Producers". 😉

In all seriousness, I think this is something important to agree on, but I think it will also require (at least) a second decision process and document if Knative wants to make a recommendation for "Event Sources" as defined here. I think "Event Importer" would probably have been a better term. Fortunately, we haven't burned the "Importer" term yet...

Reply:
Nope, I do mean "source" - assuming you're talking about this sentence:
> And in that scenario the consumer would expect things like the CE-source to be something related to the original event producer.

See: https://github.com/cloudevents/spec/pull/404 for my thoughts on this.

But, in general I view Knative's job w.r.t. event sources as utilities to help the user subscribe to event producers and to get those events delivered to their app. To the app, the fact that Kn is in the picture at all is irrelevant to the app doing its job of processing the event.  And, in fact, asking the app to know about Kn makes it harder for them to move the code some place else. As I've mentioned in some slack chats, if the event producer already generated CloudEvents then these event adapters would be more like proxies. So, in that light, the fact that they need to do "some work" to create the CloudEvent isn't material to what's going on from an app's POV - its just necessary busywork to comply with CE. The adapter is not meant to really add any semantic value beyond compliance with CE - so therefore, making the CE appear like it came from the adapter instead of the real event producer/source goes against that view.